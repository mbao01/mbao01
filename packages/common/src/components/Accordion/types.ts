import * as AccordionPrimitive from "@radix-ui/react-accordion";

export type AccordionProps = React.ComponentPropsWithoutRef<typeof AccordionPrimitive.Root>;

export type AccordionContentProps = React.ComponentPropsWithoutRef<
  typeof AccordionPrimitive.Content
>;

export type AccordionItemProps = React.ComponentPropsWithoutRef<typeof AccordionPrimitive.Item>;

export type AccordionTriggerProps = React.ComponentPropsWithoutRef<
  typeof AccordionPrimitive.Trigger
>;
