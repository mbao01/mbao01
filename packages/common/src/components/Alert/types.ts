import { type VariantProps } from "../../libs";
import { getAlertClasses } from "./constants";

export type AlertProps = React.HTMLAttributes<HTMLDivElement> &
  VariantProps<typeof getAlertClasses>;

export type AlertTitleProps = React.HTMLAttributes<HTMLHeadingElement>;

export type AlertDescriptionProps = React.HTMLAttributes<HTMLDivElement>;
