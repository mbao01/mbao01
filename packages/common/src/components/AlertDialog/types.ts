import * as AlertDialogPrimitive from "@radix-ui/react-alert-dialog";
import { type VariantProps } from "../../libs";
import { getButtonClasses } from "../Button/constants";
import {
  getDialogContentClasses,
  getDialogDescriptionClasses,
  getDialogFooterClasses,
  getDialogHeaderClasses,
  getDialogOverlayClasses,
  getDialogTitleClasses,
} from "../Dialog/constants";

export type AlertDialogProps = React.ComponentPropsWithoutRef<typeof AlertDialogPrimitive.Root>;

export type AlertDialogOverlayProps = React.ComponentPropsWithoutRef<
  typeof AlertDialogPrimitive.Overlay
> &
  VariantProps<typeof getDialogOverlayClasses>;

export type AlertDialogContentProps = React.ComponentPropsWithoutRef<
  typeof AlertDialogPrimitive.Content
> &
  VariantProps<typeof getDialogContentClasses>;

export type AlertDialogHeaderProps = React.HTMLAttributes<HTMLDivElement> &
  VariantProps<typeof getDialogHeaderClasses>;

export type AlertDialogFooterProps = React.HTMLAttributes<HTMLDivElement> &
  VariantProps<typeof getDialogFooterClasses>;

export type AlertDialogTitleProps = React.ComponentPropsWithoutRef<
  typeof AlertDialogPrimitive.Title
> &
  VariantProps<typeof getDialogTitleClasses>;

export type AlertDialogDescriptionProps = React.ComponentPropsWithoutRef<
  typeof AlertDialogPrimitive.Description
> &
  VariantProps<typeof getDialogDescriptionClasses>;

export type AlertDialogActionProps = React.ComponentPropsWithoutRef<
  typeof AlertDialogPrimitive.Cancel
> &
  VariantProps<typeof getButtonClasses>;

export type AlertDialogCancelProps = React.ComponentPropsWithoutRef<
  typeof AlertDialogPrimitive.Cancel
> &
  VariantProps<typeof getButtonClasses>;
