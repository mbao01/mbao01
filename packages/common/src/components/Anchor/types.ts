import { VariantProps } from "../../libs";
import { getAnchorClasses } from "./constant";

export type AnchorProps = React.AnchorHTMLAttributes<HTMLAnchorElement> & {
  isExternal?: boolean;
} & VariantProps<typeof getAnchorClasses>;
