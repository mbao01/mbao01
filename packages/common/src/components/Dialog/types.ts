import * as DialogPrimitive from "@radix-ui/react-dialog";
import { type VariantProps } from "../../libs";
import {
  getDialogCloseClasses,
  getDialogContentClasses,
  getDialogDescriptionClasses,
  getDialogFooterClasses,
  getDialogHeaderClasses,
  getDialogOverlayClasses,
  getDialogTitleClasses,
} from "./constants";

export type DialogProps = React.ComponentPropsWithoutRef<typeof DialogPrimitive.Root>;

export type DialogOverlayProps = React.ComponentPropsWithoutRef<typeof DialogPrimitive.Overlay> &
  VariantProps<typeof getDialogOverlayClasses>;

export type DialogCloseProps = React.ComponentPropsWithoutRef<typeof DialogPrimitive.Close> &
  VariantProps<typeof getDialogCloseClasses>;

export type DialogContentProps = React.ComponentPropsWithoutRef<typeof DialogPrimitive.Content> &
  VariantProps<typeof getDialogContentClasses> & {
    showClose?: boolean;
    closeProps?: DialogCloseProps;
  };

export type DialogHeaderProps = React.HTMLAttributes<HTMLDivElement> &
  VariantProps<typeof getDialogHeaderClasses>;

export type DialogFooterProps = React.HTMLAttributes<HTMLDivElement> &
  VariantProps<typeof getDialogFooterClasses>;

export type DialogTitleProps = React.ComponentPropsWithoutRef<typeof DialogPrimitive.Title> &
  VariantProps<typeof getDialogTitleClasses>;

export type DialogDescriptionProps = React.ComponentPropsWithoutRef<
  typeof DialogPrimitive.Description
> &
  VariantProps<typeof getDialogDescriptionClasses>;
