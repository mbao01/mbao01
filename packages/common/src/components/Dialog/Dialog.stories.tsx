import type { Meta, StoryContext, StoryFn, StoryObj } from "@storybook/react";
import { Dialog } from "./Dialog";
import { Button } from "../Button";

const withDialog = (
  _: StoryFn,
  context: StoryContext<React.ComponentPropsWithoutRef<typeof Dialog.Content>>
) => {
  return (
    <Dialog>
      <Dialog.Trigger asChild>
        <Button outline>Edit Profile</Button>
      </Dialog.Trigger>
      <Dialog.Content className="sm:max-w-[425px]">
        <Dialog.Header>
          <Dialog.Title>Edit profile</Dialog.Title>
          <Dialog.Description>
            Make changes to your profile here. Click save when you're done.
          </Dialog.Description>
        </Dialog.Header>
        Some content goes here!
        <Dialog.Footer>
          <Button type="submit">Save changes</Button>
        </Dialog.Footer>
      </Dialog.Content>
    </Dialog>
  );
};

// More on how to set up stories at: https://storybook.js.org/docs/react/writing-stories/introduction#default-export
const meta = {
  title: "Components/Dialog",
  parameters: {
    // Optional parameter to center the component in the Canvas. More info: https://storybook.js.org/docs/react/configure/story-layout
    layout: "centered",
  },
  // This component will have an automatically generated Autodocs entry: https://storybook.js.org/docs/react/writing-docs/autodocs
  tags: ["autodocs"],
  // More on argTypes: https://storybook.js.org/docs/react/api/argtypes
  argTypes: {},
  decorators: [withDialog],
} satisfies Meta<typeof Dialog.Content>;

export default meta;
type Story = StoryObj<typeof meta>;

// More on writing stories with args: https://storybook.js.org/docs/react/writing-stories/args
export const Default: Story = {
  args: {},
};
