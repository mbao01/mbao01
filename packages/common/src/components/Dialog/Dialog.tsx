"use client";

import * as React from "react";
import * as DialogPrimitive from "@radix-ui/react-dialog";
import { Cross2Icon } from "@radix-ui/react-icons";
import type {
  DialogContentProps,
  DialogDescriptionProps,
  DialogFooterProps,
  DialogHeaderProps,
  DialogOverlayProps,
  DialogProps,
  DialogTitleProps,
} from "./types";
import { cn } from "../../utilities";
import {
  getDialogCloseClasses,
  getDialogContentClasses,
  getDialogDescriptionClasses,
  getDialogFooterClasses,
  getDialogHeaderClasses,
  getDialogOverlayClasses,
  getDialogTitleClasses,
} from "./constants";

const Dialog = (props: DialogProps) => <DialogPrimitive.Root {...props} />;

const DialogOverlay = React.forwardRef<
  React.ElementRef<typeof DialogPrimitive.Overlay>,
  DialogOverlayProps
>(({ className, ...props }, ref) => (
  <DialogPrimitive.Overlay
    ref={ref}
    className={cn(getDialogOverlayClasses(), className)}
    {...props}
  />
));
DialogOverlay.displayName = DialogPrimitive.Overlay.displayName;

const DialogContent = React.forwardRef<
  React.ElementRef<typeof DialogPrimitive.Content>,
  DialogContentProps
>(
  (
    {
      className,
      children,
      closeProps,
      variant = "dialog",
      showClose = true,
      side = "right",
      ...props
    },
    ref
  ) => (
    <DialogPrimitive.Portal>
      <DialogOverlay />
      <DialogPrimitive.Content
        ref={ref}
        className={cn(
          getDialogContentClasses({
            variant,
            side: variant === "sheet" ? side : undefined,
          }),
          className
        )}
        {...props}
      >
        {children}
        {showClose ? (
          <DialogPrimitive.Close
            {...closeProps}
            className={cn(getDialogCloseClasses(), closeProps?.className)}
          >
            <Cross2Icon className="h-4 w-4" />
            <span className="sr-only">Close</span>
          </DialogPrimitive.Close>
        ) : null}
      </DialogPrimitive.Content>
    </DialogPrimitive.Portal>
  )
);
DialogContent.displayName = DialogPrimitive.Content.displayName;

const DialogHeader = ({ className, ...props }: DialogHeaderProps) => (
  <div className={cn(getDialogHeaderClasses(), className)} {...props} />
);
DialogHeader.displayName = "DialogHeader";

const DialogFooter = ({ className, ...props }: DialogFooterProps) => (
  <div className={cn(getDialogFooterClasses(), className)} {...props} />
);
DialogFooter.displayName = "DialogFooter";

const DialogTitle = React.forwardRef<
  React.ElementRef<typeof DialogPrimitive.Title>,
  DialogTitleProps
>(({ className, ...props }, ref) => (
  <DialogPrimitive.Title ref={ref} className={cn(getDialogTitleClasses(), className)} {...props} />
));
DialogTitle.displayName = DialogPrimitive.Title.displayName;

const DialogDescription = React.forwardRef<
  React.ElementRef<typeof DialogPrimitive.Description>,
  DialogDescriptionProps
>(({ className, ...props }, ref) => (
  <DialogPrimitive.Description
    ref={ref}
    className={cn(getDialogDescriptionClasses(), className)}
    {...props}
  />
));
DialogDescription.displayName = DialogPrimitive.Description.displayName;

Dialog.Trigger = DialogPrimitive.Trigger;
Dialog.Portal = DialogPrimitive.Portal;
Dialog.Close = DialogPrimitive.Close;
Dialog.Header = DialogHeader;
Dialog.Title = DialogTitle;
Dialog.Description = DialogDescription;
Dialog.Content = DialogContent;
Dialog.Footer = DialogFooter;
Dialog.Overlay = DialogOverlay;

export { Dialog };
