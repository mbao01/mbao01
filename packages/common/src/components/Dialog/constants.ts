import { cva } from "../../libs";

export const getDialogOverlayClasses = cva(
  "fixed inset-0 z-50 bg-black/80 data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0"
);

export const getDialogContentClasses = cva(
  "fixed z-50 gap-4 p-6 shadow-lg transition bg-base-100 w-full",
  {
    variants: {
      variant: {
        dialog:
          "border duration-200 left-[50%] top-[50%] grid max-w-lg translate-x-[-50%] translate-y-[-50%] data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:fade-out-0 data-[state=open]:fade-in-0 data-[state=closed]:zoom-out-95 data-[state=open]:zoom-in-95 data-[state=closed]:slide-out-to-left-1/2 data-[state=closed]:slide-out-to-top-[48%] data-[state=open]:slide-in-from-left-1/2 data-[state=open]:slide-in-from-top-[48%] sm:rounded-lg",
        sheet:
          "ease-in-out data-[state=open]:animate-in data-[state=closed]:animate-out data-[state=closed]:duration-300 data-[state=open]:duration-500",
      },
      side: {
        top: "inset-x-0 top-0 border-b data-[state=closed]:slide-out-to-top data-[state=open]:slide-in-from-top",
        bottom:
          "inset-x-0 bottom-0 border-t data-[state=closed]:slide-out-to-bottom data-[state=open]:slide-in-from-bottom",
        left: "inset-y-0 left-0 h-full w-3/4 border-r data-[state=closed]:slide-out-to-left data-[state=open]:slide-in-from-left sm:max-w-sm",
        right:
          "inset-y-0 right-0 h-full w-3/4 border-l data-[state=closed]:slide-out-to-right data-[state=open]:slide-in-from-right sm:max-w-sm",
      },
    },
    defaultVariants: {
      variant: "dialog",
    },
  }
);

export const getDialogHeaderClasses = cva("flex flex-col space-y-1.5 text-center sm:text-left");

export const getDialogFooterClasses = cva(
  "flex flex-col-reverse sm:flex-row sm:justify-end sm:space-x-2"
);

export const getDialogTitleClasses = cva("text-lg font-semibold leading-none tracking-tight");

export const getDialogDescriptionClasses = cva("text-sm text-muted-foreground");

export const getDialogCloseClasses = cva(
  "absolute right-4 top-4 rounded-xs opacity-70 ring-offset-background transition-opacity hover:opacity-100 focus:outline-hidden focus:ring-2 focus:ring-ring focus:ring-offset-2 disabled:pointer-events-none data-[state=open]:bg-accent data-[state=open]:text-muted-foreground"
);
