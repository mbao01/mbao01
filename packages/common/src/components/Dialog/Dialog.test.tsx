import { render, screen, within } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import { describe, expect, it } from "vitest";
import { Dialog } from "./Dialog";
import { type DialogContentProps } from "./types";

describe("Dialog", () => {
  const renderDialog = (contentProps?: DialogContentProps) =>
    render(
      <Dialog>
        <Dialog.Trigger>Edit Profile</Dialog.Trigger>
        <Dialog.Content className="sm:max-w-[425px]" {...contentProps}>
          <Dialog.Header>
            <Dialog.Title>Edit profile</Dialog.Title>
            <Dialog.Description>
              Make changes to your profile here. Click save when you are done.
            </Dialog.Description>
          </Dialog.Header>
          Some content goes here!
          <Dialog.Footer>Footer section</Dialog.Footer>
        </Dialog.Content>
      </Dialog>
    );

  it("renders a dialog trigger", () => {
    const { asFragment } = renderDialog();

    expect(screen.getByRole("button", { name: "Edit Profile" })).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it.each([
    { variant: "dialog", description: "dialog" },
    { variant: "sheet", description: "sheet" },
    { variant: "sheet", side: "left", description: "sheet on the left" },
    { variant: "sheet", side: "right", description: "sheet on the left" },
    { variant: "sheet", side: "top", description: "sheet on top" },
    { variant: "sheet", side: "bottom", description: "sheet on bottom" },
  ] as const)("renders a $description", async ({ variant, side }) => {
    const user = userEvent.setup();
    const { asFragment } = renderDialog({ variant, side });

    const dialogTrigger = screen.getByRole("button", { name: "Edit Profile" });

    expect(screen.queryByRole("dialog")).not.toBeInTheDocument();

    await user.click(dialogTrigger);

    const dialog = screen.getByRole("dialog");

    // dialog heading
    expect(within(dialog).getByRole("heading", { name: "Edit profile" })).toBeInTheDocument();

    // dialog description
    expect(
      within(dialog).getByText("Make changes to your profile here. Click save when you are done.")
    ).toBeInTheDocument();

    // dialog content
    expect(within(dialog).getByText("Some content goes here!")).toBeInTheDocument();

    // dialog footer
    expect(within(dialog).getByText("Footer section")).toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });

  it("should open and close dialog", async () => {
    const user = userEvent.setup();
    const { asFragment } = renderDialog();
    const dialogTrigger = screen.getByRole("button", { name: "Edit Profile" });

    expect(screen.queryByRole("dialog")).not.toBeInTheDocument();

    await user.click(dialogTrigger);

    expect(screen.queryByRole("dialog")).toBeInTheDocument();

    await user.click(screen.getByRole("button", { name: "Close" }));

    expect(screen.queryByRole("dialog")).not.toBeInTheDocument();
    expect(asFragment()).toMatchSnapshot();
  });
});
