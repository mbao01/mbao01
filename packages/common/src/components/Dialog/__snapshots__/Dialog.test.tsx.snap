// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Dialog > renders a 'dialog' 1`] = `
<DocumentFragment>
  <button
    aria-controls="radix-_r_3_"
    aria-expanded="true"
    aria-haspopup="dialog"
    data-state="open"
    type="button"
  >
    Edit Profile
  </button>
</DocumentFragment>
`;

exports[`Dialog > renders a 'sheet on bottom' 1`] = `
<DocumentFragment>
  <button
    aria-controls="radix-_r_i_"
    aria-expanded="true"
    aria-haspopup="dialog"
    data-state="open"
    type="button"
  >
    Edit Profile
  </button>
</DocumentFragment>
`;

exports[`Dialog > renders a 'sheet on the left' 1`] = `
<DocumentFragment>
  <button
    aria-controls="radix-_r_9_"
    aria-expanded="true"
    aria-haspopup="dialog"
    data-state="open"
    type="button"
  >
    Edit Profile
  </button>
</DocumentFragment>
`;

exports[`Dialog > renders a 'sheet on the left' 2`] = `
<DocumentFragment>
  <button
    aria-controls="radix-_r_c_"
    aria-expanded="true"
    aria-haspopup="dialog"
    data-state="open"
    type="button"
  >
    Edit Profile
  </button>
</DocumentFragment>
`;

exports[`Dialog > renders a 'sheet on top' 1`] = `
<DocumentFragment>
  <button
    aria-controls="radix-_r_f_"
    aria-expanded="true"
    aria-haspopup="dialog"
    data-state="open"
    type="button"
  >
    Edit Profile
  </button>
</DocumentFragment>
`;

exports[`Dialog > renders a 'sheet' 1`] = `
<DocumentFragment>
  <button
    aria-controls="radix-_r_6_"
    aria-expanded="true"
    aria-haspopup="dialog"
    data-state="open"
    type="button"
  >
    Edit Profile
  </button>
</DocumentFragment>
`;

exports[`Dialog > renders a dialog trigger 1`] = `
<DocumentFragment>
  <button
    aria-controls="radix-_r_0_"
    aria-expanded="false"
    aria-haspopup="dialog"
    data-state="closed"
    type="button"
  >
    Edit Profile
  </button>
</DocumentFragment>
`;

exports[`Dialog > should open and close dialog 1`] = `
<DocumentFragment>
  <button
    aria-controls="radix-_r_l_"
    aria-expanded="false"
    aria-haspopup="dialog"
    data-state="closed"
    type="button"
  >
    Edit Profile
  </button>
</DocumentFragment>
`;
